compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -1295,6 +1287,11 @@ private void checkWorkerPeerWorkerUsageInsideWorker(DiagnosticPos pos, BSymbol s
         }
     }
 
+    private boolean isCurrentPositionInWorker(SymbolEnv env) {
+        return env.scope.owner != null
+                && (((BInvokableSymbol) env.scope.owner).flags & Flags.WORKER) == Flags.WORKER;
```

Comment:
This check needs to be recursive. The following slight modification to the test below makes the reference to the peer worker valid.
```ballerina
    worker wx returns int {
        any a = <- wy;
        "h" -> wy;
        future<int> fi;
        var fn = function () {
            fi = wy; // should be illegal, but not.
        };
        fn();
        return wait fi;
    }
```

Reply:
Agreed, updated the code.